### PR TITLE
ICU-20036 CurrencyPluralInfo class improve handling of OOM errors

### PR DESCRIPTION
**Pull request by jefgen**
Not all code paths that allocate memory check/handle OOM errors in the CurrencyPluralInfo class.
Use an internal error code (fInternalStatus) to better report errors during clone and copy construction.

### CODE REVIEW

File: icu4c/source/i18n/currpinf.cpp [hunk 1]
```diff
@@ -246,10 +291,13 @@ CurrencyPluralInfo::setupCurrencyPluralPattern(const Locale& loc, UErrorCode& st
...
 
-    NumberingSystem *ns = NumberingSystem::createInstance(loc,status);
+    LocalPointer<NumberingSystem> ns(NumberingSystem::createInstance(loc, status), status);
+    if (U_FAILURE(status)) {
+        return;
+    }
     UErrorCode ec = U_ZERO_ERROR;
-    UResourceBundle *rb = ures_open(NULL, loc.getName(), &ec);
-    UResourceBundle *numElements = ures_getByKeyWithFallback(rb, gNumberElementsTag, NULL, &ec);
+    UResourceBundle *rb = ures_open(nullptr, loc.getName(), &ec);
```
[the remark this answers is not in the document]
Comment (jefgen) [reply icu4c/source/i18n/currpinf.cpp]: Thanks Shane, that would likely be a good idea. :)

---
In reply to: [207378377](https://github.com/unicode-org/icu/pull/17#discussion_r207378377) [](ancestors = 207378377)

File: icu4c/source/i18n/currpinf.cpp [hunk 2]
```diff
@@ -206,26 +234,43 @@ CurrencyPluralInfo::setCurrencyPluralPattern(const UnicodeString& pluralCount,
...
 
-
 void 
 CurrencyPluralInfo::initialize(const Locale& loc, UErrorCode& status) {
     if (U_FAILURE(status)) {
         return;
     }
-    delete fLocale;
-    fLocale = loc.clone();
+    if (fLocale) {
```
[the remark this answers is not in the document]
Comment (jefgen) [reply icu4c/source/i18n/currpinf.cpp]: Thanks Shane. I was trying to mimic the style within each function (if checked for `nullptr`, then I would too. If it didn't then I didn't).
However, since I'm already making a bunch of changes it would probably be better to be consistent.
Since calling delete on `nullptr` is safe, it seems like it would be cleaner to omit the checks.

---
In reply to: [207376490](https://github.com/unicode-org/icu/pull/17#discussion_r207376490) [](ancestors = 207376490)

File: icu4c/source/i18n/currpinf.cpp [hunk 3]
```diff
@@ -296,18 +285,18 @@ CurrencyPluralInfo::setupCurrencyPluralPattern(const Locale& loc, UErrorCode& st
         return;
     }
     UErrorCode ec = U_ZERO_ERROR;
-    UResourceBundle *rb = ures_open(nullptr, loc.getName(), &ec);
-    UResourceBundle *numElements = ures_getByKeyWithFallback(rb, gNumberElementsTag, nullptr, &ec);
-    rb = ures_getByKeyWithFallback(numElements, ns->getName(), rb, &ec);
-    rb = ures_getByKeyWithFallback(rb, gPatternsTag, rb, &ec);
+    LocalUResourceBundlePointer rb(ures_open(nullptr, loc.getName(), &ec));
+    LocalUResourceBundlePointer numElements(ures_getByKeyWithFallback(rb.getAlias(), gNumberElementsTag, nullptr, &ec));
+    LocalUResourceBundlePointer numSystemScript(ures_getByKeyWithFallback(numElements.getAlias(), ns->getName(), nullptr, &ec));
```
Comment (sffc) [inline icu4c/source/i18n/currpinf.cpp]: It's fine, but I think it should still be safe to use the third argument of `ures_getByKeyWithFallback` rather than making a new LocalUResourceBundlePointer for every level in the resource bundle tree.

File: icu4c/source/i18n/currpinf.cpp [hunk 4]
```diff
@@ -218,10 +215,8 @@ void
 CurrencyPluralInfo::setPluralRules(const UnicodeString& ruleDescription,
                                    UErrorCode& status) {
     if (U_SUCCESS(status)) {
-        if (fPluralRules) {
-            delete fPluralRules;
-            fPluralRules = nullptr;
-        }
+        delete fPluralRules;
+        fPluralRules = nullptr;
```
Comment (sffc) [inline icu4c/source/i18n/currpinf.cpp]: Optional: this assignment `fPluralRules = nullptr;` is redundant with the following line that just reassigns it.  Feel free to leave as-is if you think it increases readability.

File: icu4c/source/i18n/currpinf.cpp (hunk 3 above)
Comment (jefgen) [reply icu4c/source/i18n/currpinf.cpp]: I thought about that too, but then I was wondering who would call close on the ResourceBundle? (and in the correct order).
IIUC, calling `orphan()` doesn't call the close function, and if you call `getAlias()`, then the original bundle might be overwritten before the `adoptInstead()` can call close.

Ex:
 rb.adoptInstead(ures_getByKeyWithFallback(numElements.getAlias(), ns->getName(), rb.orphan(), &ec));

 rb.adoptInstead(ures_getByKeyWithFallback(numElements.getAlias(), ns->getName(), rb.getAlias(), &ec));


---
In reply to: [207658933](https://github.com/unicode-org/icu/pull/17#discussion_r207658933) [](ancestors = 207658933)

File: icu4c/source/i18n/currpinf.cpp (hunk 4 above)
Comment (jefgen) [reply icu4c/source/i18n/currpinf.cpp]: Thanks Shane. I will push another commit that address this, but I think I will leave the other LocalUResourceBundlePointers for now.

File: icu4c/source/i18n/currpinf.cpp (hunk 3 above)
Comment (sffc) [reply icu4c/source/i18n/currpinf.cpp]: What I meant was, I think it is safe to do,

```
LocalUResourceBundlePointer rb(ures_open(...));
ures_getByKeyWithFallback(rb.getAlias(), "...", rb.getAlias(), &ec);
ures_getByKeyWithFallback(rb.getAlias(), "...", rb.getAlias(), &ec);
ures_getByKeyWithFallback(rb.getAlias(), "...", rb.getAlias(), &ec);
```

Example of a place in ICU code that uses this pattern already:

https://github.com/unicode-org/icu/blob/778d0a6d1d46faa724ead19613bda84621794b72/icu4c/source/i18n/reldtfmt.cpp#L450

I'm double-checking with Markus that this is safe.

Comment (sffc) [reply icu4c/source/i18n/currpinf.cpp]: After studying [uresbund.cpp](https://github.com/unicode-org/icu/blob/778d0a6d1d46faa724ead19613bda84621794b72/icu4c/source/common/uresbund.cpp#L1212), I can't find any code paths in `ures_getByKeyWithFallback` and related functions that *obviously* return a different object than the one you passed via the fillIn parameter.  Markus didn't have an answer.  The documentation also does not provide any clues.

There is one thing that we can assume, and it's that when you pass a fillIn parameter to `ures_getByKeyWithFallback`, that you are passing ownership to that function, because most call sites, including the one you are refactoring, overwrite the original `rb` pointer with the return value.  In most cases, `ures_getByKeyWithFallback` hands ownership of fillIn right back again by returning the same pointer; the question which I can't give a 100% certain answer to is whether fillIn is *always* the return value when it is non-null.

What I can see by looking at the code is that you should be able to expect better performance by using the fillIn, because without the fillIn, you add extra mallocs.

I think the pattern I posted above is probably fine, since I can find other places in ICU code that use it, but if you are hesitant, then I think the following pattern is also definitely safe:

```
LocalUResourceBundlePointer rb(ures_open(...));
rb.adoptInstead(ures_getByKeyWithFallback(rb.getAlias(), "...", rb.orphan(), &ec));
rb.adoptInstead(ures_getByKeyWithFallback(rb.getAlias(), "...", rb.orphan(), &ec));
rb.adoptInstead(ures_getByKeyWithFallback(rb.getAlias(), "...", rb.orphan(), &ec));
```

`rb.getAlias()` gets called first, then `rb.orphan()`, then `ures_getByKeyWithFallback` does its stuff, then you save the result back into `rb`.

Comment (jefgen) [reply icu4c/source/i18n/currpinf.cpp]: Thanks Shane.
My hesitation about using that exact pattern was more about the question of "who calls the  `ures_close`" once the code is done with the Resource Bundle, in order to ensure that there are no leaks. I was thinking that when you pass the fillIn it might not close.

However, from taking a close look at the code, it looks like it will automatically call the appropriate close methods if the `fillIn` argument is not `NULL`.
Inside the `init_resb_result` function:
```c++
    if(resB == NULL) {
        // .... omitted...
    } else {
        if(resB->fData != NULL) {
            entryClose(resB->fData);
        }
        if(resB->fVersion != NULL) {
            uprv_free(resB->fVersion);
        }
        // .... omitted...
    }
```

So, in short, yes it does seem to be safe to use this pattern.
Thanks Shane.
